### DISCUSSION TITLE
Contexts and Middlewares in a separate project.

### DISCUSSION THREAD

**Discussion by danielbayley80** [trpc/trpc#3894] [Q&A]
Hello.

I'm using express and have previously created middleware in separate projects. For example, my user authentication logic / handler is in a separate module, shopping module, etc.  each has a shopmodule.intialise() which returns a handler or similar.

This allows me to compartmentalize my code and add/remove handlers. For example I have an authentication handler :

```
      requireRole(roles: string[]) {
            return async (req: Request, res: Response, next: NextFunction) => { // do auth stuff }
```

I've re-written this as follows :

```
const requireRole = (roles: string[]) => {
      return t.middleware(({ next, ctx }) => { // do auth stuff }
```

but it depends on t, which is defined once and is not available in my user auth package.

How do I abstract out my contexts and routers in to separate packages?

e.g. my user module, shop module, etc. are all separate projects and need their own contexts/handlers.

**Comment by KATT** [comment]
Maybe make the `initTRPC` root object in its own package?

Or make functions that are independent of trpc that you can use in middlewares?

**Reply by danielbayley80** [reply]
Hmmm yeah, that would do it I think, but in this case :

`const t = initTRPC.context<typeof createContext>().create();`

I would be stuck with the same context. however I then looked in to contexts in a bit more detail and saw this : 

https://trpc.io/docs/context

It appears that you can declare multiple initTRCP's.

So what I *think* might work, is having every module self contained with its own initTRCP.

Then when I call 

`app.use(userAuthService.initialize()); `

I can register trcp within the initialise method.

I haven't played with this theory but that's where I'm at with my journey.

**Reply by mmlngl** [reply]
Did it work? How'd you get on?
